plugins/PagePerformance/javascripts/PagePerformance.js

```diff
@@ -0,0 +1,136 @@
+var PagePerformance = function() {
+
+    function getDataTableFromApiMethod(apiMethod)
+    {
+        var div = $(require('piwik/UI').DataTable.getDataTableByReport(apiMethod));
+        if (div.length && div.data('uiControlObject')) {
+            return div.data('uiControlObject');
+        }
+    }
+
+    function getLabelFromTr ($tr, apiMethod) {
+        var label;
+
+        if (apiMethod && 0 === apiMethod.indexOf('Actions.')) {
+            // for now only use this for Actions... I know a hack :( Otherwise in Search Engines
+            // it would show "http://www.searchenginename.org" instead of "SearchEngineName"
+            label = $tr.attr('data-url-label');
+        }
+
+        if (!label) {
+            label = $tr.find('.label .value').text();
+        }
+
+        if (label) {
+            label = $.trim(label);
+        }
+
+        return label;
+    }
+
+
+    function getDimensionFromApiMethod(apiMethod)
+    {
+        if (!apiMethod) {
+            return;
+        }
+
+        var dataTable = getDataTableFromApiMethod(apiMethod);
+        var metadata  = getMetadataFromDataTable(dataTable);
+
+        if (metadata && metadata.dimension) {
+            return metadata.dimension;
+        }
+    }
+
+    function getMetadataFromDataTable(dataTable)
+    {
+        if (dataTable) {
+
+            return dataTable.getReportMetadata();
+        }
+    }
+
+    function findTitleOfRowHavingRawSegmentValue(apiMethod, rawSegmentValue)
+    {
+        var $tr = $('[data-report="' + apiMethod + '"] tr[data-segment-filter="' + rawSegmentValue + '"]').first();
+
+        return getLabelFromTr($tr, apiMethod);
+    }
+
+    function setPopoverTitle(apiMethod, label, index) {
+        var dataTable = getDataTableFromApiMethod(apiMethod);
+
+        if (!dataTable) {
+            if (index < 15) {
+                // this is needed when the popover is opened before the dataTable is there which can often
+                // happen when opening the popover directly via URL (broadcast.popoverHandler)
+                setTimeout(function () {
+                    setPopoverTitle(apiMethod, label, index + 1);
+                }, 150);
+            }
+            return;
+        }
+
+        var type = getDimensionFromApiMethod(apiMethod);
+
+        var separator = ' > '; // LabelFilter::SEPARATOR_RECURSIVE_LABEL
+        var labelParts = label.split(separator);
+        for (var i = 0; i < labelParts.length; i++) {
+            var labelPart = labelParts[i].replace('@', '');
+            labelParts[i] = $.trim(decodeURIComponent(labelPart));
+        }
+        var delimiter = piwik.config.action_url_category_delimiter;
+        if(apiMethod.indexOf('PageTitles') >= 0) {
+            delimiter = piwik.config.action_title_category_delimiter;
+        }
+        label = labelParts.join(delimiter);
```

Comment:
In some other places we do eg 

```js

        segmentName = piwikHelper.escape(segmentName);
        segmentName = piwikHelper.htmlEntities(segmentName);
        segmentValue = piwikHelper.escape(segmentValue);
        segmentValue = piwikHelper.htmlEntities(segmentValue);
        segmentName = segmentName.replace(/(&amp;)(#[0-9]{2,5};)/g, '&$2');
        segmentValue = segmentValue.replace(/(&amp;)(#[0-9]{2,5};)/g, '&$2');
```

Wonder if that's needed here too for the label? Considering `setTtile()` might do a `jQuery().html(title)` should the label be longer than 60 characters and maybe a user could have quotes in the title or so? Not sure... (might be good to debug or double check)

Reply:
Will add such an encoding, to be sure there can't be any kind of XSS